### ISSUE TITLE
Filter woocommerce_single_product_image_thumbnail_html needs context

### ISSUE THREAD

**Issue by anastis** [woocommerce/woocommerce#21992]
**Is your feature request related to a problem? Please describe.**
The filter `woocommerce_single_product_image_thumbnail_html` is used both in product-image.php and product-thumbnails.php but there is no solid way of determining if we're working with the main image or not.
[The suggestion](https://github.com/woocommerce/woocommerce/issues/14929#issuecomment-299852203) to compare the attachment ID with the featured image ID is flawed, as the same image can be assigned as the featured image, and be part of the gallery.

The only way to be sure that only the HTML of the main image is modified, is to override the respective template, which a) we're encouraged (by you guys) to use filters/actions instead, and b) they keep changing in minor releases, so we don't want to override them in the first place.

**Describe the solution you'd like**
A third parameter could be added, indicating whether the image is main or not. In its simplest form, it can be a boolean, but perhaps a string context could be passed instead, as the HTML currently has 3 different states (main image, main placeholder, gallery thumb).

**Describe alternatives you've considered**
Detecting whether the image is main or not, based on the passed HTML, is not a) efficient b) deterministic and c) future-proof, as addons/plugins may change the markup as well as WooCommerce may change it in future versions.

**Comment by kloon** [comment]
I think you might be confusing the name of the filter with the actual images on the screen. Take this screenshot http://cld.wthms.co/Ov8hRJ The only images passed through that filter is the ones displayed in the main big box, the smaller thumbnails at the bottom is actually not run through that filter but generated by Flexslider.

Inspect the elements and you will see what I mean. The other attachments are just hidden next to the main featured image on page load. Flexslider generates the gallery thumbnails and does the hiding and showing of the bigger thumbnails based on the gallery images clicked.

Also the way the images are queued will also have the main image be first. I don't think Mike suggested you inspect the html to find the main image, you can do a simple if statement like `if ( get_post_thumbnail_id() === $attachement_id )` to determine if you are dealing with the featured image or not, no need for inspecting html.

Perhaps if you give me an idea of what you are trying to do I can point you in the right direction with an example?

**Comment by anastis** [comment]
You are correct, assuming that the theme declares support for `wc-product-gallery-slider`.

You are not correct however, since support is optional, and if the theme doesn't support it, which is the case for a lot of themes built way before this feature was introduced (and is unfeasible or undesirable to support it), then the gallery thumbs displayed are what actually gets outputted by product-thumbnails.php.

--- start of example

So, assume a theme that *doesn't* do `add_theme_support( 'wc-product-gallery-slider' );`. Say I want to add a class on the `<a>` tag of just my thumbnails, but not on the main product image. Also assume that the end-user has added a bunch of images in the product gallery, *including* the one that s/he assigned as the main image (users will be users, they can and they will).
With that in mind:
```
add_filter( 'woocommerce_single_product_image_thumbnail_html', 'add_class_to_thumbs', 10, 2 );
function add_class_to_thumbs( $html, $attachment_id ) {
	if ( get_post_thumbnail_id() !== intval( $attachment_id ) ) {
		$html = str_replace( '<a ', '<a class="thumb-class" ', $html );
	}

	return $html;
}
```

`intval()` is necessary as `$attachment_id` is sometimes string instead of int.
The above, will add `.thumb-class` to both the main image and the gallery thumbs, which is not what's desired.

--- end of example

Some other notes:

> Also the way the images are queued will also have the main image be first.
To keep state between the runs of the hooked function is... messy (to my tastes at least). We'd need a global or static variable. Is this a better approach compared to a third parameter in the filter?

Perhaps my approach and way of thinking is completely different than your team's, so if you could briefly explain, I'd be grateful. What I can't seem to understand is, why WooCommerce treats the main image and the gallery thumbs as the same thing in terms of the `woocommerce_single_product_image_thumbnail_html` filter, but are clearly distinguished otherwise; both have their own template, and `wc_get_gallery_image_html()` has a parameter for exactly the same reason we're having this discussion.

Perhaps the solution isn't a third parameter in `woocommerce_single_product_image_thumbnail_html`, for whichever reason the team deems it's inappropriate. Perhaps `wc_get_gallery_image_html()` should be filterable instead.

**Comment by kloon** [comment]
OK, so instead of thinking of adding css classes, why not use css selectors to target the a of the children? As I mentioned, the first image will always be the primary image, with that out of the way we can use a css selector along the line of `.woocommerce-product-gallery .woocommerce-product-gallery__image:nth-child(n+2)` which will target all the divs except the primary image.

I will work on something, but this will only be going out in the next minor release ie 3.6 and will not solve any issues for you with customers running older versions of WooCommerce.

**Comment by kloon** [comment]
See #22005 for new filter.

**Comment by anastis** [comment]
It's not just about css classes.
In one theme, I needed to add data attributes for in the main image. In another theme, in just the thumbnails, I had to remove the (recently introduced) div that wraps the images.

All these are currently doable, it just requires either modifying a template, or detecting if we're working with the main image using unreliable methods.

It's not about my customers. I've already provided them new theme versions that work with the latest WooCommerce. What I'm suggesting is about making the developers' lives easier, and improve WooCommerce itself.
